### PR TITLE
API: allow step!=1 slice with IntervalIndex

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

See discussion https://github.com/pandas-dev/pandas/pull/16386#discussion_r374807689

cc @jorisvandenbossche @jschendel

### CODE REVIEW

**Comment by jreback** [comment]
I suppose should add a whatsnew bug fix entry.

File: doc/source/whatsnew/v1.1.0.rst [hunk 1]
```diff
@@ -145,7 +145,7 @@ Strings
 
 Interval
 ^^^^^^^^
-
+- Bug in :class:`IntervalIndex` slicing that prevented slicing with ``step > 1`` (:issue:`31658`)
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v1.1.0.rst]: I would put this in the "other enhancements" section

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]
Is there a test that ensures that slicing with an interval as step still raises an error?

File: pandas/tests/indexing/interval/test_interval_new.py [hunk 2]
```diff
@@ -143,9 +144,27 @@ def test_loc_with_slices(self):
...
+    def test_slice_interval_step(self):
+        # GH#31658 allows for integer step!=1, not Interval step
+        s = self.s
+        msg = (
+            "cannot do slice indexing on "
+            "<class 'pandas.core.indexes.interval.IntervalIndex'> with "
+            "these indexers .* of <class 'pandas._libs.interval.Interval'>"
+        )
+        with pytest.raises(TypeError, match=msg):
+            s[0 : 4 : Interval(0, 1)]
```
Comment (jorisvandenbossche) [inline pandas/tests/indexing/interval/test_interval_new.py]: I would test it where start and stop are also intervals. 
And can you move it to the `test_loc_with_slices` a bit above that is testing slicing with Intervals ?

File: pandas/tests/indexing/interval/test_interval_new.py [hunk 3]
```diff
@@ -143,9 +144,27 @@ def test_loc_with_slices(self):
...
+        result2 = s[0:4][::2]
+        tm.assert_series_equal(result2, expected)
+
+    def test_slice_interval_step(self):
+        # GH#31658 allows for integer step!=1, not Interval step
+        s = self.s
+        msg = (
+            "cannot do slice indexing on "
+            "<class 'pandas.core.indexes.interval.IntervalIndex'> with "
+            "these indexers .* of <class 'pandas._libs.interval.Interval'>"
```
Comment (jorisvandenbossche) [inline pandas/tests/indexing/interval/test_interval_new.py]: The current error message I get is more comprehensible I think (with `s` from the tests here):

```
In [41]: s[Interval(3, 4):Interval(4, 5):Interval(1, 2)]                                                                                                                                                           
...
ValueError: cannot support not-default step in a slice

In [42]: s[1:2:Interval(1, 2)]                                                                                                                                                           
...
ValueError: cannot support not-default step in a slice
```

File: pandas/tests/indexing/interval/test_interval_new.py [hunk 4]
```diff
@@ -128,6 +128,7 @@ def test_loc_with_slices(self):
         with pytest.raises(NotImplementedError, match=msg):
             s[Interval(3, 4, closed="left") :]
 
+        # FIXME: dont leave commented-out
```
Comment (jorisvandenbossche) [inline pandas/tests/indexing/interval/test_interval_new.py]: I think you can just remove it, or turn it in a test. It's the exact same error and message as the tests just above with other non-existing intervals.

File: doc/source/whatsnew/v1.1.0.rst [hunk 5]
```diff
@@ -42,6 +42,7 @@ Other enhancements
 ^^^^^^^^^^^^^^^^^^
 
 - :class:`Styler` may now render CSS more efficiently where multiple cells have the same styling (:issue:`30876`)
+- Bug in :class:`IntervalIndex` slicing that prevented slicing with ``step > 1`` (:issue:`31658`)
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v1.1.0.rst]: Can you word it as an enhancement now?

**Review by jorisvandenbossche — COMMENTED** [review]
@jbrockmendel what would you expect for this:

```
s = Series(np.arange(5), IntervalIndex.from_breaks(np.arange(2, 14, 2))) 
s[1.5:9.5:2]
```

I think that is less clear, which might be the reason it was disallowed initially.

**Comment by jbrockmendel** [comment]
> what would you expect for [...] `s[1.5:9.5:2]` [...] I think that is less clear, which might be the reason it was disallowed initially.

I agree that is less obvious, as that is not positional.

File: pandas/tests/indexing/interval/test_interval_new.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/tests/indexing/interval/test_interval_new.py]: > start and stop are also intervals

out of scope

>  And can you move it to the test_loc_with_slices a bit above that is testing slicing with Intervals ?

sure

Comment (jorisvandenbossche) [reply pandas/tests/indexing/interval/test_interval_new.py]: > > start and stop are also intervals
> 
> out of scope

I think that is not out of scope, as that hits this path, and right now actually ensured there is a proper error message.

**Comment by jorisvandenbossche** [comment]
> I agree that is less obvious, as that is not positional.

So let's not allow that?

**Comment by jbrockmendel** [comment]
> So let's not allow that?

OK I see, I thought that the float-slice would have been disallowed somewhere else, but evidently not.  So yes, this needs to be updated to restore that.

**Comment by jbrockmendel** [comment]
@jorisvandenbossche double-checking: suppose we get integer sllce that is valid for either loc or iloc:

```
import pandas as pd
import numpy as np

ser = pd.Series(np.arange(5), pd.IntervalIndex.from_breaks(np.arange(2, 14, 2))) 
key = slice(3, 6)

ser[key]
ser.loc[key]
ser.iloc[key]

key2 = 3
ser[key2]
ser.loc[key2]
ser.iloc[key2]
```

For the slice we treat the ambiguous `__getitem__` key as positional, while for the int we treat it as a label.  That's on purpose, right?

**Comment by jbrockmendel** [comment]
Closing to clear the queue (in particular #31786 should happen before this), will revisit

**Comment by jreback** [comment]
can you rebase. look ok to me. @jschendel

**Comment by jbrockmendel** [comment]
rebased+green

**Comment by jreback** [comment]
can you rebase.

**Comment by jbrockmendel** [comment]
rebased+green

File: pandas/core/indexers.py [hunk 6]
```diff
@@ -20,6 +21,34 @@
 # Indexer Identification
 
 
+def is_valid_positional_slice(slc: slice) -> bool:
+    """
+    Check if a slice object can be interpretd as a positional indexer.
```
Comment (jschendel) [inline pandas/core/indexers.py]: nitpick: interpretd --> interpreted

**Review by jschendel — APPROVED** [review]
lgtm, thanks!

File: pandas/core/indexers.py (hunk 6 above)
Comment (jbrockmendel) [reply pandas/core/indexers.py]: updated

**Comment by jreback** [comment]
thanks 2 followups:

1) @jorisvandenbossche comment about having start/stop as intervals, pls create an issue
2) does this need an example in the docs?

**Pull request merged**